### PR TITLE
[MoE] Make structured fused grouped-expert projection the default

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.18.0) (oldest at bottom):
* #4543
* #4542
* #4541
* #4665
* #4623
* #4559
* #4652
* #4591
* #4592
* #4540
* #4539
* __->__ #4643

Summary:
Carry forward the reviewed implementation from pytorch/torchtitan#4552 while
storing routed-expert W13 as `[E, 2, F, D]`. The explicit gate/up axis matches
DistMoE's native parameter layout and keeps the feature axis independently
shardable. Standard grouped GEMM consumes a zero-copy `[E, 2F, D]` view and
returns `[R, 2, F]` before the existing activation.

Logical initialization and checkpoint keys remain `w1_EFD`, `w2_EDF`, and
`w3_EFD`. Routing, activation, grouped-GEMM, and dispatcher behavior are
unchanged. The physical fusion intentionally changes seeded random
initialization order, so the affected real-PG integration goldens are refreshed;
checkpoint-loaded weights retain their logical values.

Fused W13 also adds a second native matrix-batch dimension. DistMuon now accepts
native `[..., rows, columns]` parameters, matching the existing Newton-Schulz
implementation, and its expert redistribution test exercises the production
`[E, 2, F, D]` layout without adding another test case.

This is the Sanket-owned replacement for #4552; Animesh's PR and stack remain
untouched.

Test Plan:
- `python -m pytest -q tests/unit_tests/gpu/flex_shard/test_dist_muon.py` (2 passed)
- `python -m pytest -q tests/unit_tests/gpu/test_fsdp_moe_sharding.py` (4 passed)
- `CUDA_VISIBLE_DEVICES='' python -m pytest -q tests/unit_tests/cpu/test_moe.py tests/unit_tests/cpu/test_inference_moe.py tests/unit_tests/cpu/test_quantization.py torchtitan/experiments/transformers_modeling_backend/tests/test_moe_parallelism.py tests/unit_tests/gpu/test_fused_swiglu_override.py -k 'not TestFusedSwiGLUGroupedExpertsNumerics and not TestFusedSwiGLUOverrideKernels'` (69 passed, 3 skipped, 6 deselected)
- TorchTitan formatting, flake8, pydoclint, codespell, and link checks passed; Pyrefly reports one existing unrelated optional-`av` diagnostic in `torchtitan/hf_datasets/multimodal/utils/video.py`.
- The real-PG A10G model suite produced the refreshed 10-step loss and gradient-norm goldens.

### CODE REVIEW

File: torchtitan/models/common/moe.py
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:1]: looks great in general

suspicious as it didn't touch
- https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/fsdp.py#L296
- not sure if we need to adapt anything here https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/models/vllm_wrapper.py#L564

**Review by tianyu-l — CHANGES_REQUESTED** [review]

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/common/moe.py
Comment (sanketpurandare) [reply torchtitan/models/common/moe.py:1]: Updated the fallback FSDP placement for the structured parameter: when the E/FSDP degree exceeds the expert count, `w13_E2FD` now shards its hidden dimension with `Shard(2)`, while `w2_EDF` retains `Shard(1)`. The distributed tests now assert both placements in the EP=1 and EP>1 paths instead of checking only the first expert parameter. The placement map is derived from the actual expert parameters, so GPT-OSS keeps its existing model-specific layout; a new four-rank GPT-OSS regression test exercises that exact compatibility path. The vLLM state-dict path now also accounts for the removed projection axis: selecting gate or up from physical `[E, 2, F, D]` translates physical `S(2)` to logical `S(1)` for each exported `[E, F, D]` tensor. The helper preserves the other mesh-axis types and rejects selecting a sharded dimension. `test_state_dict_layouts_include_split_expert_weights` now asserts distinct physical and exported layouts. The full 8-GPU real-PG model and feature integration suites passed with these changes.

File: torchtitan/distributed/flex_shard/dist_muon.py
Comment (tianyu-l) [inline torchtitan/distributed/flex_shard/dist_muon.py:1]: @shuhuayu @weifengpy could you review the flex_shard changes

File: torchtitan/distributed/fsdp.py [hunk 1]
```diff
@@ -326,14 +325,18 @@ def apply_fsdp_to_decoder(
                     efsdp_ep_size *= dp_storage_mesh["cp"].size()
 
             if efsdp_ep_size > num_experts:
-                expert_shard_placement = Shard(1)
+                expert_param_placements = {
+                    param: Shard(2 if name == "w13_E2FD" else 1)
+                    for name, param in experts.named_parameters()
+                }
```
Comment (tianyu-l) [inline torchtitan/distributed/fsdp.py:331]: I assume this explicit term will be refactored away in next PR when introducing `GroupedLinear` so that you can apply something similar https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/fsdp.py#L299

File: torchtitan/models/deepseek_v3/state_dict_adapter.py [hunk 2]
```diff
@@ -157,7 +157,9 @@ def to_hf(self, state_dict: dict[str, Any]) -> dict[str, Any]:
         1. Convert between the HF shape and the torchtitan shape.
         2. Split the GroupedExperts' weight into separate expert's weight.
         """
-        state_dict = self._native_fused_linears_to_hf(state_dict)
+        state_dict = self._to_logical_expert_state(
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/state_dict_adapter.py:160]: I feel we could merge this ability into the `_native_fused_linears_to_hf` function, maybe in next PR?

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -59,16 +69,30 @@ class Config(Module.Config):
...
+            )
         )
         self.w2_EDF = nn.Parameter(
             torch.empty(config.num_experts, config.dim, config.hidden_dim)
         )
-        self.w3_EFD = nn.Parameter(
-            torch.empty(config.num_experts, config.hidden_dim, config.dim)
-        )
         self.activation_fn = config.activation_fn.build()
+        self.register_load_state_dict_pre_hook(self._merge_w13_on_load)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: save hook and load hook should happen in a pair -- a single merge hook doesn't make sense?

File: torchtitan/rl/model/vllm_wrapper.py [hunk 4]
```diff
@@ -562,13 +562,6 @@ def get_state_dict_layouts(self) -> dict[str, SpmdType]:
                 for state_name, layout in sharding_config.state_shardings.items():
                     layouts[f"{module_prefix}{state_name}"] = layout
 
-                # Fused grouped experts expose split gate/up state-dict keys
-                # while the layout is declared on the fused w13 parameter.
-                w13_layout = sharding_config.state_shardings.get("w13")
-                if w13_layout is not None:
-                    for state_name in ("w1_EFD", "w3_EFD"):
-                        layouts[f"{module_prefix}{state_name}"] = w13_layout
-
```
Comment (tianyu-l) [inline torchtitan/rl/model/vllm_wrapper.py:571]: I thought this can be removed only when
- we don't have save/load hooks in model (i.e. MoE)
- we move the state dict fuse / unfuse to hf state_dict_adapters

In this PR you kept the merge hook in model, which I haven't thought about what would happen. But I suggest we be consistent on load / save.

File: torchtitan/models/kimi_k3/__init__.py [hunk 5]
```diff
@@ -260,11 +261,13 @@ def _latent_moe_config(
                 hidden_dim=expert_hidden_dim,
                 num_experts=num_experts,
                 activation_fn=SiTUGLU.Config(beta=4.0, linear_beta=25.0),
-                param_init={
-                    "w1_EFD": partial(nn.init.trunc_normal_, std=0.02),
-                    "w2_EDF": partial(nn.init.trunc_normal_, std=0.02),
-                    "w3_EFD": partial(nn.init.trunc_normal_, std=0.02),
-                },
+                param_init=fused_grouped_experts_param_init(
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k3/__init__.py]: used `w13_E2FD`, but the muon pattern still uses `w3_EFD` and `w1_EFD`, can you revise that part?

**Review by shuhuayu — DISMISSED** [review]

File: torchtitan/distributed/flex_shard/dist_muon.py
Comment (shuhuayu) [reply torchtitan/distributed/flex_shard/dist_muon.py]: flex_shard related changes lgtm.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/distributed/fsdp.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/distributed/fsdp.py]: Implemented in the child PR #4539. `_linear_param_shard_placements()` now handles both stacked `Linear` and `GroupedLinear` modules by deriving the matrix-row dimension from each parameter rank, including ordinary `[E, F, D]` grouped weights when E/FSDP must row-shard. No W13/W2 FQN special case remains.

File: torchtitan/models/deepseek_v3/state_dict_adapter.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/models/deepseek_v3/state_dict_adapter.py]: Implemented in #4539. The existing fused-linear state-adapter helpers now accept an explicit routed-W13 opt-in and own both split and merge for HF schemas with separate expert projections. GPT-OSS and Qwen3.5 continue to use their fused HF layouts.

File: torchtitan/rl/model/vllm_wrapper.py (hunk 4 above)
Comment (sanketpurandare) [reply torchtitan/rl/model/vllm_wrapper.py]: Removed the model-level load hook. Native model, DCP, optimizer, and EMA state now consistently use the physical `w13_E2FD` key, while the HF adapters own the paired split/merge conversion at the external checkpoint boundary.

File: torchtitan/models/kimi_k3/__init__.py (hunk 5 above)
Comment (sanketpurandare) [reply torchtitan/models/kimi_k3/__init__.py]: Updated Kimi K3's DistMuon `expert_projections` to the physical `w13_E2FD` and `w2_EDF` parameters. The focused optimizer-config test now checks both compute-layout FQNs, verifies that the DistMuon regex selects both native parameters, and rejects the stale `w1_EFD`/`w3_EFD` names.

Comment (shuhuayu) [reply torchtitan/models/kimi_k3/__init__.py]: cool, thx.

**Review by shuhuayu — APPROVED** [review]

**Comment by weifengpy** [comment]
merge to main？
